src/main/java/io/vertx/ext/mail/impl/SMTPConnectionPool.java

```diff
@@ -17,243 +17,216 @@
 
   SMTPConnectionPool(Vertx vertx, MailConfig config) {
+    this.vertx = vertx;
     this.config = config;
-    maxSockets = config.getMaxPoolSize();
-    keepAlive = config.isKeepAlive();
-    this.prng = new PRNG(vertx);
-    this.authOperationFactory = new AuthOperationFactory(prng);
-
     // If the hostname verification isn't set yet, but we are configured to use SSL, update that now
     String verification = config.getHostnameVerificationAlgorithm();
     if ((verification == null || verification.isEmpty()) && !config.isTrustAll() &&
-        (config.isSsl() || config.getStarttls() != StartTLSOptions.DISABLED)) {
+      (config.isSsl() || config.getStarttls() != StartTLSOptions.DISABLED)) {
       // we can use HTTPS verification, which matches the requirements for SMTPS
       config.setHostnameVerificationAlgorithm("HTTPS");
     }
-
     netClient = vertx.createNetClient(config);
+    this.prng = new PRNG(vertx);
+    this.authOperationFactory = new AuthOperationFactory(prng);
+    if (config.getPoolCleanerPeriod() > 0 && config.isKeepAlive() && config.getKeepAliveTimeout() > 0) {
+      timerID = vertx.setTimer(config.getPoolCleanerPeriod(), this::checkExpired);
+    }
+  }
+
+  private void checkExpired(long timer) {
+    getSMTPEndPoint().checkExpired(ar -> {
+      if (ar.succeeded()) {
+        ar.result().forEach(c -> c.quitCloseConnection(Promise.promise()));
+      }
+    });
+    synchronized (this) {
+      if (!closed) {
+        timerID = vertx.setTimer(config.getPoolCleanerPeriod(), this::checkExpired);
+      }
+    }
   }
 
   AuthOperationFactory getAuthOperationFactory() {
     return authOperationFactory;
   }
 
   void getConnection(String hostname, Handler<AsyncResult<SMTPConnection>> resultHandler) {
-    log.debug("getConnection()");
-    this.hostname = hostname;
-    if (closed) {
-      resultHandler.handle(Future.failedFuture("connection pool is closed"));
-    } else {
-      getConnection0(resultHandler);
-    }
+    getConnection(hostname, vertx.getOrCreateContext(), resultHandler);
   }
 
-  void close() {
-    close(null);
+  void getConnection(String hostname, Context ctx, Handler<AsyncResult<SMTPConnection>> resultHandler) {
+    getConnection0(hostname, ctx, resultHandler, 0);
   }
 
-  synchronized void close(Handler<Void> finishedHandler) {
-    if (closed) {
-      throw new IllegalStateException("pool is already closed");
-    } else {
-      closed = true;
-      closeFinishedHandler = finishedHandler;
-      closeAllConnections();
-      this.prng.close();
+  private void getConnection0(String hostname, Context ctx, Handler<AsyncResult<SMTPConnection>> resultHandler, final int i) {
+    synchronized (this) {
+      if (closed) {
+        resultHandler.handle(Future.failedFuture("connection pool is closed"));
+        return;
+      }
     }
+    ContextInternal contextInternal = (ContextInternal)ctx;
+    Promise<Lease<SMTPConnection>> promise = contextInternal.promise();
```

Comment:
could this use Future#transform instead of onComplete ?

Reply:
Updated the futures chain to get a connection using `recover` and `transform`